src/Crypto/Hash/PoseidonNative.hs

```diff
@@ -0,0 +1,3229 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+
+module Crypto.Hash.PoseidonNative (poseidon) where
+
+import Data.List(foldl')
+import qualified Data.Primitive.Array as Array
+import qualified Data.Primitive.SmallArray as SmallArray
+
+modulus :: Integer
+modulus = 21888242871839275222246405745257275088548364400416034343698204186575808495617
+
+mulmod :: Integer -> Integer -> Integer
+mulmod a b = (a * b) `mod` modulus
+
+summod :: Integer -> Integer -> Integer
+summod a b = (a + b) `mod` modulus
+
+sig :: Integer -> Integer
+sig inVal =
+  let in2 = mulmod inVal inVal
+      in4 = mulmod in2 in2
+  in mulmod in4 inVal
+
+ark :: Integer -> Integer -> Integer
+ark inVal cc = summod inVal cc
+
+data NineRings =
+  NineRings !Integer !Integer !Integer !Integer !Integer !Integer !Integer !Integer !Integer
+
+-- the real constructor for NineRings
+tabulateRingsTo :: Int -> (Int -> Integer) -> NineRings
+tabulateRingsTo n f =
+    NineRings
+      (ifGTE 0)
+      (ifGTE 1)
+      (ifGTE 2)
+      (ifGTE 3)
+      (ifGTE 4)
+      (ifGTE 5)
+      (ifGTE 6)
+      (ifGTE 7)
+      (ifGTE 8)
+    where
+    ifGTE k = if n >= k then f k else 0
+{-# inline conlike tabulateRingsTo #-}
+
+-- without inlining this and `ca`, NineRings is never unpacked into arguments
+mix :: Int -> NineRings -> Int -> Integer
+mix nInputs (NineRings in1 in2 in3 in4 in5 in6 in7 in8 in9) i =
+  foldl'
+    (\acc (k, ring) -> summod acc (mulmod ring (getMatrix (nInputs - 1) i k)))
+    0
+    ([0..nInputs] `zip` [in1,in2,in3,in4,in5,in6,in7,in8,in9])
+{-# inline mix #-}
+
+ca :: Int -> Int -> NineRings -> Int -> Int -> Integer
+ca nInputs t nr j i = ark (mix nInputs nr j) (getC (nInputs - 1) (j + (t * i)))
+{-# inline ca #-}
+
+poseidonWithRounds :: Int -> Int -> [Integer] -> Integer
+poseidonWithRounds nRoundsF nRoundsP inputs =
+    let
+        -- initial state
+        ms = tabulateRingsTo nInputs $ \n ->
+          sig $
+            if n == 0 then ark 0 (getC (nInputs - 1) 0)
+            else ark (inputs !! (n - 1)) (getC (nInputs - 1) n)
+
+        -- TODO: gas each round
+        rounds = (nRoundsF + nRoundsP) - 1
+
+        -- iterative rounds
+        outRings =
+          foldl' applyRound ms [1..rounds]
+
+        -- return
+        out = mix nInputs outRings 0
+    in
+        out
+    where
+        nInputs = length inputs
+        shouldApplySig roundNumber =
+          (roundNumber < nRoundsF `div` 2) || (roundNumber >= nRoundsP + nRoundsF `div` 2)
```

Comment:
Are `roundNumber`s and `nRoundsF` every negative? If so, you can avoid some unncessary checks here with `quot` (also, looking for when `rem` would apply instead of `mod` is also a good idea).

Reply:
I don't believe they are. Good catch.

Reply:
Great point, we hadn't even tried this. Unbelievably though, doing this replacement here, or in the `mulmod` or `summod` functions, actually slows things down a bit. Very bizarre...

Reply:
That's quite weird actually o.0